datastream-server/src/main/java/com/linkedin/datastream/server/assignment/StickyPartitionAssignmentStrategy.java

```diff
@@ -190,15 +211,18 @@ public StickyPartitionAssignmentStrategy(Optional<Integer> maxTasks, Optional<In
         if (dg.belongsTo(task)) {
           Set<String> toMovePartitions = new HashSet<>(task.getPartitionsV2());
           toMovePartitions.retainAll(allToReassignPartitions);
-          confirmedPartitionsTaskMap.put(task.getDatastreamTaskName(), toMovePartitions);
-          toMovePartitions.forEach(p -> partitionToSourceTaskMap.put(p, task.getDatastreamTaskName()));
+          if (!toMovePartitions.isEmpty()) {
```

Comment:
I feel we can avoid splitting this in two steps. Please check and let me know your opinion.

``` // We first confirm that the partitions in the target assignment which can be removed, and we find out its source task
    // If the partitions cannot be found from any task, we ignore these partitions
    currentAssignment.keySet().forEach(instance -> {
      Set<DatastreamTask> tasks = currentAssignment.get(instance);
      Set<String> partitionsAcrossAllDatastreamTasks = currentAssignment.get(instance).stream().filter(dg::belongsTo)
          .map(DatastreamTask::getPartitionsV2).flatMap(Collection::stream).collect(Collectors.toSet());
      //Remove the partitions that do not need migration.
      allToReassignPartitions.removeAll(partitionsAcrossAllDatastreamTasks);
      tasks.forEach(task -> {
        if (dg.belongsTo(task)) {
          Set<String> toMovePartitions = new HashSet<>(task.getPartitionsV2());
          toMovePartitions.retainAll(allToReassignPartitions);
          if (!toMovePartitions.isEmpty()) {
            // Only update the confirmedPartitionsTaskMap if a partition is indeed being deleted from it
            confirmedPartitionsTaskMap.put(task.getDatastreamTaskName(), toMovePartitions);
            toMovePartitions.forEach(p -> partitionToSourceTaskMap.put(p, (DatastreamTaskImpl) task));
          }
        }
      });
    });`

Reply:
I tried this as is, but it doesn't work (`preprocessedTargetAssignment` is used in a later step, so I also added code here to create this too) and it's also getting very tricky to get it to work. Just removing the `partitionsAcrossAllDatastreamTasks` from `allToReassignPartitions` results in `allToReassignPartitions` becoming  empty eventually. I tried creating a copy of `allToReassignPartitions` too, and removed from that, but  the logic wasn't correct. Tried to retain all of `partitionsAcrossAllDatastreamTasks`, then there are issues with creating the `confirmedPartitionsTaskMap`, and when I try to fix that then I see that some partitions are completely removed from the final assignment, etc.

At least for me, this code is quite tricky to reason with, so the two step assignment just feels cleaner and easier to understand. It may be okay to perform a two step process here, since it's just recalculating the assignments. What do you think?

Reply:
Found a way to make this work. Silly me should have tried moving the allToReassignPartitions to within the targetAssignment.containsKey() block.

Reply:
Can you please check this code block? This logic is correct (with one minor addition required) but not so easy to get it. 
```
 if (targetAssignment.containsKey(instance)) {
        // Process the targetAssignment to filter out partitions that are supposed to move to the target instance that
        // they are already on to avoid unnecessary partition movements. That is, if partition1 is currently on
        // instance1 and is supposed to move to instance1, remove it from the targetAssignment as this should be a
        // no-op. Accordingly the allToReassignPartitions should also be updated to remove such partitions.
        //Partitions not suppose to move
        Set<String> partitionsRemovedFromTargetAssignment = new HashSet<>(targetAssignment.get(instance));
        partitionsRemovedFromTargetAssignment.retainAll(partitionsAcrossAllDatastreamTasks);

        Set<String> updatedTargetPartitionList = new HashSet<>(targetAssignment.get(instance));
        updatedTargetPartitionList.removeAll(partitionsRemovedFromTargetAssignment);
        // filter invalid partitions
        updatedTargetPartitionList.retainAll(allToReassignPartitions);

        processedTargetAssignment.computeIfAbsent(instance,
            val -> updatedTargetPartitionList.isEmpty() ? null : updatedTargetPartitionList);

        allToReassignPartitions.removeAll(partitionsRemovedFromTargetAssignment);
      }`

In this, I have reordered your calls,
a. first find the partitions  from the target, which are not moving. 
b. Remove these partitions from the targetAssignment and retain  only the valid partitions in allToReassignPartitions (missing in your code) and save it in updatedTargetPartitionList
c. Remove these partitions from allToReassignPartitions

Reply:
done.

> b. Remove these partitions from the targetAssignment and retain  only the valid partitions in allToReassignPartitions (missing in your code) and save it in updatedTargetPartitionList

That step isn't really necessary (though I added it) since allToReassignPartitions has already filtered out invalid partitions, and all the other data-structures are created based on this.